### ISSUE TITLE
Serializing HibernateProxy objects

### ISSUE THREAD

**Issue by mantithetical** [javers/javers#91]
Attempting to serialize HibernateProxy objects throws an error like:

 CLASS_NOT_MANAGED JaVers runtime error - given javaClass 'mypackage.myclass_$$_javassist_171'

I have tried to add a HibernateProxySerializer like so:

    private static class HibernateProxySerializer extends JsonTypeAdapterTemplate<HibernateProxy> {
		@Inject
		private TypeMapper typeMapper;

		@Override
		public HibernateProxy fromJson(JsonElement json,
				JsonDeserializationContext jsonDeserializationContext) {
			return jsonDeserializationContext.deserialize(json, typeMapper.getDehydratedType(HibernateProxy.class));
		}

		@Override
		public JsonElement toJson(HibernateProxy sourceValue,
				JsonSerializationContext jsonSerializationContext) {
			Object deProxied = sourceValue.getHibernateLazyInitializer().getImplementation();
			return jsonSerializationContext.serialize(deProxied, typeMapper.getDehydratedType(HibernateProxy.class));
		}

		@Override
		public Class getValueType() {
			return HibernateProxy.class;
		}
	}

And then registered the serializer like so:

	javers = JaversBuilder.javers().registerJaversRepository(mongoRepo)
		.registerValueTypeAdapter(new HibernateProxySerializer()).build();

However, I am still running into the same error.

**Comment by bartoszwalacik** [comment]
Hi mantithetical 
two things:

1. Could you please come up with running example which reproduces this error, and push it to github?

2. Don't think serializing Hibernate proxies is a right way. Proxies are runtime-generated wrappers for domain objects. JaVers should have a feature to skip/ignore proxies

**Comment by mantithetical** [comment]
1. Let me back to you with a compact example
2. Ignoring proxies is a good start. However, an option to unproxy the proxies before serialization could also be useful.

**Comment by bartoszwalacik** [comment]
as far as I remember, unproxing Hibernate proxies is not trivial. Hibernate proxies domain objetcs 
and all collections.
We don't want to create dependency to Hibernate in javers-core, so unproxying will be implemented as another javers module or generic approach.

First of all we need runnable example for testing, could you push to github compact project with Hibernate/H2 setup?

**Comment by mantithetical** [comment]
Here's an example https://github.com/mantithetical/javers-hibernate-proxy. The call to javers is in PodcastDaoJPA2Impl. Gson allows one to skip classes; perhaps, we could exploit that feature to ignore proxies. 

There's another issue with circular references when serializing bi-directional mappings; I will open a separate issue for that. Thanks for your help.

**Comment by bartoszwalacik** [comment]
got it, I will start to working on unproxing issue at the weekend

**Comment by bartoszwalacik** [comment]
hi, I've prepared some unit test but I can't access your repo to push it, could you give me write access ?

**Comment by mantithetical** [comment]
Done.

**Comment by bartoszwalacik** [comment]
I've pushed unfinished unit test
https://github.com/mantithetical/javers-hibernate-proxy/blob/master/src/test/java/org/codingpedia/demo/rest/dao/impl/JaversIntegrationTest.java

I'm not in JPA/Hibernate for few yeras, could you help me to configure this test?
I need autowired entityManeger with H2 database

**Comment by mantithetical** [comment]
Sure, I will take a stab at it shortly.

**Comment by bartoszwalacik** [comment]
hi, any progress in configuring our test?

**Comment by mantithetical** [comment]
Sorry about the delay; I was traveling. Let me get back to you.

**Comment by bartoszwalacik** [comment]
No hurry , just checking if this issue is still alive. I think, this feature would be valuable.

**Comment by mantithetical** [comment]
Done. You should now see the following error when you run that test.
```java
JaversException: ENTITY_INSTANCE_WITH_NULL_ID JaVers runtime error - Found Entity instance of class 'org.codingpedia.demo.rest.entities.Edition_$$_jvstd69_1' with null id
```

**Comment by mantithetical** [comment]
Hey there, any progress on this? Is there anything I could help with?

**Comment by bgalek** [comment]
I'm researching this problem right now ;) Need to have one two free evening and I'll make pull request :) thx for beeing ready to help :+1:

**Comment by bartoszwalacik** [comment]
@bgalek when come up with working version, we can release it as snapshot and get feedback from @mantithetical

**Comment by bgalek** [comment]
to be released at the end of May :)

**Comment by mantithetical** [comment]
Amazing!

**Comment by phelypp** [comment]
Hello Friends,

We really want to use Javers however need this issue
Any position to delivery date?

thank you

**Comment by bartoszwalacik** [comment]
@bgalek is working on it, i cant push him :)

**Comment by phelypp** [comment]
Hello Javers :)

You have a definite date?
But we will make an outline in our application.

Thank you

**Comment by bgalek** [comment]
Working on it right now, hope to write here about results later today :)

**Comment by bgalek** [comment]
I've just updated branch with this feature, hope @bartoszwalacik @pawelszymczyk will check it out soon and maybe we could release ASAP :)

**Comment by phelypp** [comment]
Good, I'm waiting merge

Thanks :)

**Comment by bartoszwalacik** [comment]
released in 1.2.11
